lib/bolt/fiber_executor.rb

```diff
@@ -0,0 +1,168 @@
+# frozen_string_literal: true
+
+require 'bolt/logger'
+require 'bolt/plan_future'
+
+module Bolt
+  class FiberExecutor
+    attr_reader :plan_futures
+
+    def initialize
+      @logger = Bolt::Logger.logger(self)
+      @id = 0
+      @plan_futures = []
+    end
+
+    # Whether there is more than one fiber running in parallel.
+    #
+    def in_parallel
+      plan_futures.length > 1
+    end
+
+    # Creates a new Puppet scope from the current Plan scope so that variables
+    # can be used inside the block and won't interact with the outer scope.
+    # Then creates a new Fiber to execute the block, wraps the Fiber in a
+    # Bolt::PlanFuture, and returns the Bolt::PlanFuture. Because the main plan
+    # doesn't need the scoping and is called by the plan compiler in PAL, this
+    # is separate from `add_future` which adds the future to the list of
+    # futures to round_robin on. This is so that the main plan can create it's
+    # own PlanFuture and add it, while all other PlanFutures are created here.
+    #
+    def create_future(scope, block, name: nil, object: nil)
```

Comment:
I talked with Alex about it, since each Future has a unique ID and users might not know what futures have been created we decided that Futures sharing a name was ok (just not sharing an ID). I'm a little on the fence about it myself, but does that seem reasonable to you?

Reply:
Ya, fine with me. We should remove it from the acceptance criteria in that case.

Reply:
Good call, removed